[sival,kmac] Add kmac_endianess test

nasahlpa opened the pull request:
This commit adds a chip level test to test the endianess feature of KMAC. This includes the message as well as the state register.

The following configurations are covered:
- Little endian message, little endian digest.
- Little endian message, big endian digest.
- Big endian message, little endian digest.
- Big endian message, big endian digest.

Currently, only the KMAC mode is covered. However, as the big/ little endian feature should be independent of the underlying mode, this should be sufficient.

vogelpi commented on sw/device/tests/kmac_endianess_test.c:
```diff
@@ -0,0 +1,207 @@
+      .entropy_seed = {0xb153e3fe, 0x09596819, 0x3e85a6e8, 0xb6dcdaba,
+                       0x50dc409c, 0x11e1ebd1},
+      .entropy_fast_process = kDifToggleEnabled,
+      .output_big_endian = true,
```
Should you also set the `message_big_endian` to true? You're using the big endian messages for this test but I think if you don't configure the corresponding struct member to true, it might actually default to false.

vogelpi commented on sw/device/tests/kmac_endianess_test.c:
```diff
@@ -0,0 +1,207 @@
+ * This test runs a KMAC operation twice, in the first run with
+ * message_big_endian true and the input message is in big ending. In the second
+ * run, the message_big_endian config is set to false and the input message is
+ * in little endian. In the end, the output digest should match.
```
It probably makes sense to also document which digest endianess you're using here.

vogelpi reviewed (commented):
Thanks @nasahlpa for working on this. The test implementation looks good to me. I have just some minor comments.

One more question: The test plan says that we should test all combinations of message and digest endianess. Assuming that you explicitly set the message endianess for the digest test to big endian, we test the following cases:
- digest test:
  - msg big, digest little
  - msg big, digest big
- msg test:
  - msg big, digest big
  - msg little, digest big

So we don't test  the combination msg little and digest little. Can you change the test to cover all combinations?

nasahlpa commented:
Thanks for the review, @vogelpi!

I tested this endianess configuration as internally the core anyway always operates on little endian and produces little endian output. In the big endian message case, the input is converted. In the little endian message case, the output is converted. So my reasoning was that testing all combinations shouldn't be necessary.

However, I see your point that we should test whether this behavior really is given, so I changed the test to cover all possible combinations. This also simplified the test :-)

vogelpi reviewed (approved):
This looks good, thanks @nasahlpa !
